packages/contracts/test/plugin/plugin-setup-processor.ts

```diff
@@ -941,68 +1146,777 @@ describe('Plugin Setup Processor', function () {
+        });
+
+        it('sets the V1 to V2 permissions', async () => {
+          expect(permissionsV1V2).to.deep.equal(
+            mockPermissionsOperations(1, 2, Operation.Grant).map(perm =>
+              Object.values(perm)
+            )
           );
+        });
 
-        await psp.applyInstallation(
-          targetDao.address,
-          pluginSetupV1,
-          pluginSetupMockRepoAddress,
-          plugin,
-          prepareInstallpermissions
-        );
+        it('updates to V3', async () => {
+          await updateHelper(
+            psp,
+            targetDao,
+            proxy,
+            repoUUPS,
+            helpersV2,
+            setupV2,
+            setupV3
+          );
+        });
+
+        it('cannot update back to V1', async () => {
+          await expect(
+            updateHelper(
+              psp,
+              targetDao,
+              proxy,
+              repoUUPS,
+              helpersV2,
+              setupV2,
+              setupV1
+            )
+          ).to.be.reverted;
+        });
+
+        it('cannot update again to V2', async () => {
+          await expect(
+            updateHelper(
+              psp,
+              targetDao,
+              proxy,
+              repoUUPS,
+              helpersV2,
+              setupV2,
+              setupV2
+            )
+          ).to.be.reverted;
+        });
+
+        context(`and updated to V3`, function () {
+          let helpersV3: string[];
+          let permissionsV2V3: PermissionOperation[];
+          let initDataV2V3: BytesLike;
+
+          beforeEach(async () => {
+            ({
+              updatedHelpers: helpersV3,
+              permissions: permissionsV2V3,
+              initData: initDataV2V3,
+            } = await updateHelper(
+              psp,
+              targetDao,
+              proxy,
+              repoUUPS,
+              helpersV2,
+              setupV2,
+              setupV3
+            ));
+          });
+
+          it('points to the V3 implementation', async () => {
+            expect(
+              await PluginV3.attach(proxy).callStatic.getImplementationAddress()
+            ).to.equal(await setupV3.callStatic.getImplementationAddress());
+          });
+
+          it('initializes the members', async () => {
+            expect(await PluginV3.attach(proxy).v1()).to.equal(1);
+            expect(await PluginV3.attach(proxy).v2()).to.equal(2);
+            expect(await PluginV3.attach(proxy).v3()).to.equal(3);
+          });
+
+          it('sets the V3 helpers', async () => {
+            expect(helpersV3).to.deep.equal(mockHelpers(3));
+          });
+
+          it('sets the V2 to V3 permissions', async () => {
+            expect(permissionsV2V3).to.deep.equal(
+              mockPermissionsOperations(2, 3, Operation.Grant).map(perm =>
+                Object.values(perm)
+              )
+            );
+          });
+
+          it('cannot update back to V1', async () => {
```

Comment:
```suggestion
          it('cannot downgrade to V1', async () => {
```
just a suggestion

Reply:
I prefer "update back" (because the function is called update).